lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -32,26 +32,186 @@
 using namespace swift;
 using namespace DerivedConformance;
 
-/// Common preconditions for Equatable and Hashable.
-static bool canDeriveConformance(NominalTypeDecl *type) {
-  // The type must be an enum.
-  // TODO: Structs with Equatable/Hashable/Comparable members
-  auto enumDecl = dyn_cast<EnumDecl>(type);
-  if (!enumDecl)
-    return false;
-
-  // The enum must have cases.
-  if (!enumDecl->hasCases())
-    return false;
-
-  // The enum must not have associated values.
-  // TODO: Enums with Equatable/Hashable/Comparable payloads
-  if (!enumDecl->hasOnlyCasesWithoutAssociatedValues())
-    return false;
-  
+/// Returns true if the given type conforms to the protocol.
+/// \p type The type whose conformance should be checked.
+/// \p protocol The protocol being requested.
+/// \return True if the type conforms to the protocol.
+static bool typeConformsToProtocol(TypeChecker &tc, DeclContext *context,
```

Comment:
I feel like this utility does not carry its weight. Note that an Optional value already has a conversion to bool, so you can just call 'if (tc.conformsToProtocol(...))', I think.

Reply:
Removed.